core/unit_test/TestBitManipulationBuiltins.hpp

```diff
@@ -425,3 +427,174 @@ TEST(TEST_CATEGORY, bit_manip_bit_width) {
   test_bit_manip_bit_width<unsigned long>();
   test_bit_manip_bit_width<unsigned long long>();
 }
+
+#undef TEST_BIT_MANIP_FUNCTION
+
+#define DEFINE_BIT_ROTATE_FUNCTION_EVAL(FUNC)                \
+  struct BitRotateFunction_##FUNC {                          \
+    template <class T>                                       \
+    static KOKKOS_FUNCTION auto eval_constexpr(T x, int s) { \
+      return Kokkos::FUNC(x, s);                             \
+    }                                                        \
+    template <class T>                                       \
+    static KOKKOS_FUNCTION auto eval_builtin(T x, int s) {   \
+      return Kokkos::Experimental::FUNC##_builtin(x, s);     \
+    }                                                        \
+    static char const* name() { return #FUNC; }              \
+  }
+
+DEFINE_BIT_ROTATE_FUNCTION_EVAL(rotl);
+DEFINE_BIT_ROTATE_FUNCTION_EVAL(rotr);
+
+#undef DEFINE_BIT_ROTATE_FUNCTION_EVAL
+
+template <class T>
+struct P {
+  using type = T;
+  T x;
+  int s;
+};
+
+template <class Space, class Func, class Arg, std::size_t N>
+struct TestBitRotateFunction {
+  Arg val_[N];
+  TestBitRotateFunction(const Arg (&val)[N]) {
+    std::copy(val, val + N, val_);
+    run();
+  }
+  void run() const {
+    int errors = 0;
+    Kokkos::parallel_reduce(Kokkos::RangePolicy<Space>(0, N), *this, errors);
+    ASSERT_EQ(errors, 0) << "Failed check no error for " << Func::name() << "("
+                         << type_helper<typename Arg::type>::name() << ", int)";
+  }
+  KOKKOS_FUNCTION void operator()(int i, int& e) const {
+    if (Func::eval_builtin(val_[i].x, val_[i].s) !=
+        Func::eval_constexpr(val_[i].x, val_[i].s)) {
+      ++e;
+      KOKKOS_IMPL_DO_NOT_USE_PRINTF(
+          "value at %x rotated by %d which is %x was expected to be %x\n",
```

Comment:
Does the statement above clearly convey the expected condition?

Reply:
It prints on failure what the arguments were `value at <x> rotated by <s>`, what the function returned `is <returned>`, and what the expected value was `was expected to be <reference>`.

What is missing?

Reply:
Nothing (in terms of expected vs. actual) is missing, but this formulation (or some variation thereof) is clearer:  "It prints on failure what the arguments were value at <x> rotated by <s>, what the function returned is <returned>, and what the expected value was was expected to be <reference>."

Reply:
I don't understand your suggestion.  Can you suggest a diff?